index.bs

```diff
@@ -4648,71 +4656,154 @@ The <dfn export for=commands>browsingContext.setViewport</dfn> command modifies
    </dd>
 </dl>
 
+<div algorithm>
+To <dfn>set device pixel ratio override</dfn> given |navigable| and |device pixel ratio|:
+
+1. Let |navigable| be the [=/navigable=] whose [=navigable/active document=] is
+   |navigable|'s [=navigable/active document=].
+
+1. If |device pixel ratio| is not null:
+
+   1. When the [=select an image source from a source set=] are run, act as if
+      the implementation's pixel density was set to |device pixel ratio| when selecting an image.
+
+   1. For the purposes of the [=resolution media feature=], act as if
+      the implementation's resolution is |device pixel ratio| dppx scaled by the page zoom.
+
+   1. [=map/Set=] [=device pixel ratio overrides=][|navigable|] to |device pixel ratio|.
+
+      Note: This will take an effect because of the patch of [[#patchs-determine-the-device-pixel-ratio]].
+
+1. Otherwise:
+
+   1. When the [=select an image source from a source set=] steps are run, use the implementation's default behavior,
+      without any changes made by previous invocations of these steps.
+
+   1. For the purposes of the [=resolution media feature=], use the implementation's default behavior,
+      without any changes made by previous invocations of these steps.
+
+   1. [=map/Remove=] |navigable| from [=device pixel ratio overrides=].
+
+1. Run [=evaluate media queries and report changes=] for [=/document=] currently loaded
+   in a specified |navigable|.
+
+</div>
+
+
+<div algorithm>
+To <dfn>set viewport</dfn> given |navigable| and |viewport|:
+
+1. If |viewport| is not null, set the width of |navigable|'s [=layout
+   viewport=] to be the <code>width</code> of |viewport| in CSS pixels and
+   set the height of the |navigable|'s [=layout viewport=] to be the
+   <code>height</code> of |viewport| in CSS pixels.
+
+1. Otherwise, set the |navigable|'s [=layout viewport=] to the
+   implementation-defined default.
+
+</div>
+
+<div algorithm="set viewport when a top-level traversable is created">
+
+When the [=set up a window environment settings object=]
```

Comment:
It's probably a wrong place to hook on, but since `WebDriver BiDi navigable created` should happen when everything is set up, and likely we have to set up viewport even before preload scripts run we need something else here, and will have to add something to html spec. So this is more of placeholder for now.

Reply:
Do we actually need to do this once per document, or only really once per navigable? It seems like ideally we run this once per top-level traversable, since we seem to skip over everything else, and then you can assert that the input navigable is a top-level traversable, and skip a bunch of checks below.

I think for now it would be better to have a sketch of under what conditions we want to run these steps (maybe "after creating a document in a new top-level traversable, before running the 'Run WebDriver BiDi preload scripts' steps") rather than being more specific in a way that's misleading.